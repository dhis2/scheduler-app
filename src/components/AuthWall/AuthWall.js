import React, { useEffect } from 'react'
import { connect } from 'react-redux'
import { node, func, string, bool } from 'prop-types'
import { CircularLoader } from '@dhis2/ui-core'
import { getMe } from '../../rootSelectors'
import { actions, selectors } from '../../data/me'
import { AbsoluteCenter } from '../AbsoluteCenter'

export const UnconnectedAuthWall = ({
    children,
    didFetch,
    errorMessage,
    isAuthorized,
    fetchMeIfNeeded,
}) => {
    useEffect(() => {
        fetchMeIfNeeded()
    }, [fetchMeIfNeeded])

    if (!didFetch) {
        return (
            <AbsoluteCenter vertical>
                <CircularLoader />
                Checking permissions
            </AbsoluteCenter>
        )
    }

    if (errorMessage) {
        return <span>{errorMessage}</span>
    }

    if (!isAuthorized) {
        return <span>You are not authorized</span>
    }

    return <React.Fragment>{children}</React.Fragment>
}

UnconnectedAuthWall.propTypes = {
    children: node.isRequired,
    didFetch: bool.isRequired,
    errorMessage: string.isRequired,
    isAuthorized: bool.isRequired,
    fetchMeIfNeeded: func.isRequired,
}

const mapStateToProps = state => {
    /* istanbul ignore next */
    const me = getMe(state)

    /* istanbul ignore next */
    return {
        didFetch: selectors.getDidFetch(me),
        errorMessage: selectors.getErrorMessage(me),
        isAuthorized: selectors.getIsAuthorized(me),
    }
}

const mapDispatchToProps = {
    fetchMeIfNeeded: actions.fetchMeIfNeeded,
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(UnconnectedAuthWall)
