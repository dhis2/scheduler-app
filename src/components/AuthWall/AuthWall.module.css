.noticeBoxWrapper {
    max-inline-size: 800px;
    margin: 0 auto;
}
