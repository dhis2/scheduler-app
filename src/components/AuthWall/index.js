export { default as AuthWall } from './AuthWall'
