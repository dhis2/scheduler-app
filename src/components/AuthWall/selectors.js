export const getAuthorized = me => {
    const { authorities } = me

    if (!authorities) {
        return false
    }

    const isAuthorized =
        authorities.includes('ALL') ||
        authorities.includes('F_SCHEDULING_ADMIN')

    return isAuthorized
}
