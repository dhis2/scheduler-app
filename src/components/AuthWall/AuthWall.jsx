import React from 'react'
import PropTypes from 'prop-types'
import { NoticeBox } from '@dhis2/ui'
import i18n from '@dhis2/d2-i18n'
import { useDataQuery } from '@dhis2/app-runtime'
import { Spinner } from '../Spinner'
import { getAuthorized } from './selectors'
import styles from './AuthWall.module.css'

const query = {
    me: {
        resource: 'me',
    },
}

const AuthWall = ({ children }) => {
    const { loading, error, data } = useDataQuery(query)

    if (loading) {
        return <Spinner />
    }

    if (error) {
        return (
            <div className={styles.noticeBoxWrapper}>
                <NoticeBox error title={i18n.t('Something went wrong')}>
                    {i18n.t(
                        'Something went wrong whilst retrieving user permissions.'
                    )}
                </NoticeBox>
            </div>
        )
    }

    const isAuthorized = getAuthorized(data.me)

    if (!isAuthorized) {
        return (
            <div className={styles.noticeBoxWrapper}>
                <NoticeBox error title={i18n.t('Not authorized')}>
                    {i18n.t(
                        "You don't have access to the Job Scheduler. Contact a system administrator to request access."
                    )}
                </NoticeBox>
            </div>
        )
    }

    return <React.Fragment>{children}</React.Fragment>
}

const { node } = PropTypes

AuthWall.propTypes = {
    children: node.isRequired,
}

export default AuthWall
