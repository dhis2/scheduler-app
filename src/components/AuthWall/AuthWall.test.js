import React from 'react'
import { shallow } from 'enzyme'
import { UnconnectedAuthWall as AuthWall } from './AuthWall'

const defaultProps = {
    children: 'Children',
    isFetching: false,
    errorMessage: '',
    isAuthorized: false,
    fetchMeIfNeeded: () => {},
}

describe('<AuthWall>', () => {
    it('renders a spinner when loading', () => {
        const props = { ...defaultProps, isFetching: true }
        const wrapper = shallow(<AuthWall {...props} />)

        expect(wrapper).toMatchSnapshot()
    })

    it('renders errors if they occur', () => {
        const props = { ...defaultProps, errorMessage: 'Something went wrong' }
        const wrapper = shallow(<AuthWall {...props} />)

        expect(wrapper).toMatchSnapshot()
    })

    it('renders a message for unauthorized users', () => {
        const props = { ...defaultProps, isAuthorized: false }
        const wrapper = shallow(<AuthWall {...props} />)

        expect(wrapper).toMatchSnapshot()
    })

    it('renders the children for authorized users', () => {
        const props = { ...defaultProps, isAuthorized: true }
        const wrapper = shallow(<AuthWall {...props} />)

        expect(wrapper).toMatchSnapshot()
    })

    it('fetches data on mount', () => {
        const props = { ...defaultProps, fetchMeIfNeeded: jest.fn() }
        const wrapper = shallow(<AuthWall {...props} />)

        expect(props.fetchMeIfNeeded).toHaveBeenCalled()
    })
})
