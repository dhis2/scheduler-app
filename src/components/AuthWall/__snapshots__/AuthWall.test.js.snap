// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthWall> renders a message for unauthorized users 1`] = `
<span>
  You are not authorized
</span>
`;

exports[`<AuthWall> renders a spinner when loading 1`] = `<CircularLoader />`;

exports[`<AuthWall> renders errors if they occur 1`] = `
<span>
  Something went wrong
</span>
`;

exports[`<AuthWall> renders the children for authorized users 1`] = `
<Fragment>
  Children
</Fragment>
`;
